tasmota/tasmota_xdrv_driver/xdrv_16_tuyamcu_v1.ino

```diff
@@ -1262,150 +1261,173 @@ void TuyaInit(void) {
-          MqttPublishPrefixTopic_P(RESULT_OR_TELE, PSTR(D_JSON_TUYA_MCU_RECEIVED));
-        } else {
-          AddLog(LOG_LEVEL_DEBUG, ResponseData());
-        }
+    }
+    else if (Tuya.byte_counter == 1) {
+      if (serial_in_byte == 0xAA) { // Only packets with header 0x55AA are valid
+        Tuya.buffer[Tuya.byte_counter++] = 0xAA;
+        Tuya.cmd_checksum = 0xFF;
       } else {
-        AddLog(LOG_LEVEL_DEBUG, ResponseData());
+        Tuya.byte_counter = 0; // if not received 0xAA right after the 0x55, reset the state machine
+        AddLog(LOG_LEVEL_DEBUG_MORE,PSTR("TYA: 0x55 without 0xAA - resync"));
       }
-      XdrvRulesProcess(0);
-
-      if (dpId != 0 && Settings->tuyamcu_topic) { // Publish a /STAT Topic ready to use for any home automation system
-        if (!Tuya.SuspendTopic) {
-          char scommand[13];
-          snprintf_P(scommand, sizeof(scommand), PSTR("DpType%uId%u"), dpDataType, dpId);
-          if (dpDataType != 3 && dpDataType != 5) { Response_P(PSTR("%u"), DataVal); }
-          else { Response_P(PSTR("%s"), DataStr); }
-          MqttPublishPrefixTopic_P(STAT, scommand);
-        }
+    }
+    else if (Tuya.byte_counter < 6) {
+      Tuya.buffer[Tuya.byte_counter++] = serial_in_byte;
+      Tuya.cmd_checksum += serial_in_byte;
+      if (Tuya.byte_counter == 6) { 
+        // Get length of data, max buffer is 256 bytes, so only taking into account lowest byte of length
+        Tuya.data_len = serial_in_byte + 6;
       }
-
-      if (!Tuya.low_power_mode) {
-        TuyaNormalPowerModePacketProcess();
+    }
+    else if (Tuya.byte_counter == Tuya.data_len) {
+      Tuya.buffer[Tuya.byte_counter++] = serial_in_byte;
+      if (Tuya.cmd_checksum == serial_in_byte) { // Compare checksum and process packet
+        AddLogBuffer(LOG_LEVEL_DEBUG_MORE,(uint8_t*)Tuya.buffer,Tuya.byte_counter);
+        TuyaProcessMessage();
       } else {
-        TuyaLowPowerModePacketProcess();
+        AddLog(LOG_LEVEL_DEBUG_MORE,PSTR("TYA: checksum error: 0x%02X instead of 0x%02X"), serial_in_byte, Tuya.cmd_checksum);
       }
-
       Tuya.byte_counter = 0;
-      Tuya.cmd_status = 0;
-      Tuya.cmd_checksum = 0;
-      Tuya.data_len = 0;
-    }                                                    // read additional packets from TUYA
+    }
     else if (Tuya.byte_counter < TUYA_BUFFER_SIZE -1) {  // add char to string if it still fits
       Tuya.buffer[Tuya.byte_counter++] = serial_in_byte;
       Tuya.cmd_checksum += serial_in_byte;
-    } else {
+    } 
+    else { // buffer overflow, erset the state machine
       Tuya.byte_counter = 0;
-      Tuya.cmd_status = 0;
-      Tuya.cmd_checksum = 0;
-      Tuya.data_len = 0;
     }
   }
+  // reset the state machine if no bytes received since a long time
+  if (Tuya.byte_counter > 0 && (millis() - time_last_byte_received) > TUYA_CMD_TIMEOUT) {
+     Tuya.byte_counter = 0;
+     time_last_byte_received = millis();
```

Comment:
I think this is not necessary. After a timeout and a new received byte the `time_last_byte_received`variable will be set in line 1384. That should be enough.

Reply:
Sounds correct.
I'm removing it